Account for change in UTC offset when performing next schedule calculations

timkpaine opened the pull request:
Closes: #7999

When performing a next value calculation using a cron schedule with a non-fixed interval (e.g. `* 1,2,3 * * *`), changes in timezone are not properly accounted for. The functions `_get_next` and `_get_prev` do not account for the delta between the `current` and `scheduled` values' respective timezones. This PR fixes that, and accordingly fixes the closed but unfixed #7999 issue.

Note that this may want to be considered a breaking change, even though it is a fix in the behavior.

timkpaine committed:
fixes #7999 - account for change in UTC offset when performing next schedule calculations

timkpaine committed:
Add newsfragment explaining change

fix lint

fix copypasta

potiuk commented on airflow/timetables/_cron.py:
```diff
@@ -113,7 +119,13 @@ def _get_prev(self, current: DateTime) -> DateTime:
+        current_offset = current.in_timezone(self._timezone).utcoffset()
+        scheduled_offset = instance(scheduled, self._timezone).utcoffset()
+        utc_offset_delta = current_offset - scheduled_offset
+        delta = naive - scheduled
```
```suggestion
```

potiuk replied:
(was repeated it seems)

potiuk reviewed (approved):
LGTM (wiht the small nit - @uranusjr - I think second opinion would be great.

timkpaine committed:
add test for leaving DST

timkpaine committed:
undo DST adjusting, reverts behavior back to the original

timkpaine committed:
update news fragment

uranusjr commented on newsfragments/30083.significant.rst:
```diff
@@ -0,0 +1,8 @@
+
+As an example, if operating in the `America/New_York` timezone when entering daylight savings time, the above schedule would result in a `10am` execution on Monday morning, instead of the previous `9am`.
+
+This functionality is now deprecated and reverted, so timezone-aware timetables will now run at the same time, e.g. a `9am` timetable will always run at `9am` local time.
```
A deprecation would imply the functionality is still available but disfavoured. This is not the case here.

timkpaine committed:
tweak language

uranusjr committed:
Revise news fragment grammar

uranusjr committed:
Remove _should_fix_dst entirely

This is private interface anyway so we don't need to keep it around.
Also move the logic into DAG.is_fixed_time_schedule() so it is not
broken by the removal of the logic.

uranusjr reviewed (approved):
I added a couple of commits to refine English grammar of the news fragment, and fix `is_fixed_time_schedule` that was broken by change.

timkpaine committed:
Properly account for transitioning across DST bounds

timkpaine commented on tests/models/test_dag.py:
```diff
@@ -677,87 +677,176 @@ def test_following_previous_schedule(self):
         """
         Make sure DST transitions are properly observed
         """
+        # First pass, start on pre side
```
starting here, I go through 2 sequences based on CET->CEST and CEST->CET. In both cases, I assert the behavior of getting next/previous from before and after the time change juncture. Please take care to review the behavior.

Taragolis commented on airflow/utils/timezone.py:
```diff
@@ -74,6 +77,18 @@ def utc_epoch() -> dt.datetime:
     return result
 
 
+def in_timezone(datetime: DateTime, tz: Timezone, dst_rule: str = pendulum.PRE_TRANSITION) -> DateTime:
```
pendulum literals PRE_TRANSITION, PRE_TRANSITION will be removed into the pendulum 3 in favor of of [`datetime.datetime.fold`](https://docs.python.org/3/library/datetime.html#datetime.datetime.fold) maybe we could use fold here, and use `dst_rule` depends on fold value `0` or `1`

Seems like PRE_TRANSITION use same behaviour as fold=0 (default in datetime.datetime) and  POST_TRANSITION same as fold=1, but better to check in [pendulum 2.1.2 code](https://github.com/sdispater/pendulum/blob/2.1.2/pendulum/tz/timezone.py#L92)

timkpaine replied:
Yeah they're the same, I can look to change this, the necessary functionality involves switching between pre and post based on which direction you're moving and which DST transition it is, to account for doing the extra hour both pre and post transition

Taragolis replied:
Yeah my point that in some point in the future (I hope near) pendulum 3 released and we need to migrate and most I guess we need to support two major versions. And if it possible to use `fold` as a parameter to functions now then transition would be less painful then. ❤️

uranusjr replied:
Pendulum 2.1.2 actually already supports using `fold` to control transition.

https://github.com/sdispater/pendulum/blob/d9aa101fe76852b75cba14d686ede923641c5b4d/pendulum/tz/timezone.py#L100-L103

This was added 6 years ago. I didn’t check which version. It should be safe to use it.

https://github.com/sdispater/pendulum/commit/5be0094a89d629af5fd2d87164d1c0b8e2c8bde8

timkpaine replied:
@uranusjr thanks for looking into this, i will make the switch

dstandish commented on airflow/utils/timezone.py:
```diff
@@ -74,6 +77,18 @@ def utc_epoch() -> dt.datetime:
+    :param tz: pendulum Timezone
+    :param dst_rule: pendulum.PRE_TRANSITION or pendulum.POST_TRANSITION
+    :return: localized datetime in timezone
+    """
```
```suggestion

    :meta private:
    """
```

dstandish committed:
Update airflow/utils/timezone.py

dstandish commented on newsfragments/30083.significant.rst:
```diff
@@ -0,0 +1,13 @@
+Non-fixed cron timetables no longer account for changing in timezone.
```
Is there a more descriptive term than "non-fixed"?  That doesn't have meaning for me, but maybe I'm alone on that.  At same time, I am not sure how to describe what it is.  Maybe "non-fixed-interval" or "not-constant-interval" or, instead of the double negative, "variable-interval"

dstandish replied:
```suggestion
Variable-interval cron timetables no longer account for DST transition
```

also, is "account for DST transition" more accurate than "account for changing in timezone"?  when you say "changing in timezone" it sounds like you mean "changing the timezone" -- like, changing the timezone of the dag?

dstandish commented on newsfragments/30083.significant.rst:
```diff
@@ -0,0 +1,13 @@
+Non-fixed cron timetables no longer account for changing in timezone.
+
+Given a cron expression resulting in an interval-based schedule, such as
+``0 9-16 * * *``, job after a timezone change (e.g. entering or exiting daylight
+saving time) used to be scheduled with an offset to the change in timezone. For
+example, when operating in *America/New_York*, the above schedule would result
+in a run at 10am on Monday morning after entering DST, instead of the previous
+9am.
+
+This was intentional to account for the change in timezone introduced in Airflow
+2.2, but caused confusion to several use cases. The functionality is now
+reverted to its pre-2.2 state, and timetables will now always run at the same
+*clock time*, e.g. 9am in the above example.
```
It seems we should state that this was a bug that we are fixing, not simply a change that we are reverting.  We have observed that it's not simply that the schedule doesn't run at the correct clock time, but that the dag actually gets stuck and fails to run.

Additionally I'm a bit confused by the 9am / 10am example.  Are you saying that the behavior before this fix would be that the dag would run at 10am until DST transitions back, 6 months later? Or only the first run?

dstandish commented on tests/models/test_dag.py:
```diff
@@ -677,87 +677,176 @@ def test_following_previous_schedule(self):
-        prev_local = local_tz.convert(prev)
+    #     prev = dag.previous_schedule(_next)
+    #     prev_local = local_tz.convert(prev)
 
```
seems something needs to be addressed here.  Should these tests be removed?  Or uncommented?  Or do we have a decision to make about what the behavior should be?

uranusjr replied:
This test actually almost passes, only fails in the final `previous_schedule` call. Since `previous_schedule` is deprecated anyway, I think it is OK to just remove that part.

I’m also converting it into a timetable test to not rely on `following_schedule`, which is also deprecated.

dstandish commented on tests/timetables/test_interval_timetable.py:
```diff
@@ -253,3 +253,83 @@ def test_cron_next_dagrun_info_alignment(last_data_interval: DataInterval, expec
+    assert cron_interval._get_prev(last_run).timestamp() == cron_fixed._get_prev(last_run).timestamp()
+
+
+def test_cron_interval_dst_next_leaving():
```
we should probably verify the DST transition behavior with the `CronTriggerTimetable` class also.  perhaps parameterizing would be an easy way to do that.

uranusjr committed:
Remove in_timezone() to rely on datetime.fold

Pendulum 3 will remove the dst_rule flag altogether, so this changes the
code to use the fold flag instead, which will be the only way moving
forward, and has been supported in Pendulum 2 for a long time anyway.

uranusjr committed:
Better wording in changelog

Co-authored-by: Daniel Standish <15932138+dstandish@users.noreply.github.com>

uranusjr committed:
Allow make_aware caller to explicitly disable fold

timkpaine committed:
Uncomment second set of tests

timkpaine commented on tests/timetables/test_interval_timetable.py:
```diff
@@ -253,3 +253,83 @@ def test_cron_next_dagrun_info_alignment(last_data_interval: DataInterval, expec
+        "0 9-16 * * 1-5", timezone=pendulum.timezone("America/New_York")
+    )
+    # Fixed cron
+    cron_fixed = CronDataIntervalTimetable("0 9 * * 1-5", timezone=pendulum.timezone("America/New_York"))
```
This and the test below need to get rid of 1-5 so that they're fixed not interval
